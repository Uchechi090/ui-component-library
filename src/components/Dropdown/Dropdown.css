.dropdown-container {
  width: 250px;
  position: relative;
}

.dropdown-list {
  border: 1px solid #d0d5dd;
  width: 100%;
  box-shadow: 0px 1px 2px rgba(16, 24, 40, 0.05);
  border-radius: 8px;
  position: absolute;
  display: none;
}

.dropdown-list.visible {
  display: block;
}

.dropdown-item {
  height: 36px;
  display: flex;
  align-items: center;
  padding-left: 10px;
  cursor: pointer;
  background-color: white;
  font-family: "Neucha", cursive;
}

.dropdown-item:hover {
  background-color: #c53678;
  color: #fff;
}

.dropdown-item:first-of-type,
.dropdown-item:first-of-type:hover {
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.dropdown-item:last-of-type,
.dropdown-item:last-of-type:hover {
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
}

.input-container {
  position: relative;
  width: 100%;
}

.dropdown-container input {
  width: 100%;
  padding: 0;
  border-width: 1px;
  height: 36px;
  padding-left: 10px;
  box-sizing: border-box;
  position: relative;
}

.input-arrow-container {
  position: absolute;
  right: 10px;
  top: 70%;
  transform: translateY(-50%);
  margin-top: -2px;
  cursor: pointer;
}

.input-arrow {
  color: #667085;
}
