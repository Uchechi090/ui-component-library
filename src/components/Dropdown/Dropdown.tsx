import React, { useState, useEffect, useRef, ChangeEvent } from "react";
import { FaAngleDown } from "react-icons/fa6";

import "./Dropdown.css";

interface DropdownProps {
  options: string[];
  selectedOption: string;
  onChange: React.Dispatch<React.SetStateAction<string>>;
}

const Dropdown: React.FC<DropdownProps> = ({
  options,
  selectedOption,
  onChange,
}) => {
  const [open, setOpen] = useState(false);
  const [inputValue, setInputValue] = useState(selectedOption);

  const containerRef = useRef<any>(null);

  useEffect(() => {
    const handleClickOutside = (event: { target: any }) => {
      if (!containerRef?.current?.contains(event.target)) {
        setOpen(false);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
  }, [containerRef]);

  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
    onChange(e.target.value);
  };

  const onInputClick = () => {
    setOpen(!open);
  };

  const onOptionSelected = (option: string) => {
    onChange !== undefined && onChange(option);
    onChange !== undefined && setInputValue(option);
    setOpen(false);
  };

  return (
    <div data-testid="dropdown-container" className="dropdown-container" ref={containerRef}>
      <div data-testid="input-container" className="input-container" onClick={onInputClick}>
        <input
          data-testid="input"
          type="text"
          value={inputValue}
          placeholder={"Select an option"}
          onChange={onInputChange}
        />
        <div className="input-arrow-container">
          <FaAngleDown className="input-arrow" />
        </div>
      </div>
      {open ? (
        <div data-testid="dropdown-list" className={`dropdown-list ${open ? "visible" : ""}`}>
          {options.map((item) => (
            <div
              data-testid="dropdown-item"
              className="dropdown-item"
              key={item}
              onClick={() => onOptionSelected(item)}
            >
              {item}
            </div>
          ))}
        </div>
      ) : null}
    </div>
  );
};

export default Dropdown;
