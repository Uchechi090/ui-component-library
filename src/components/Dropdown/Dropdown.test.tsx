import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Dropdown from "./Dropdown";

describe("<Dropdown />", () => {
  const mockChangeValue = jest.fn();
  const options = ["Apple", "Banana", "Carrot", "Dates"];
  let selectedOption = options[1];

  it("should show props", () => {
    const { getByTestId, asFragment } = render(
      <Dropdown
        options={options}
        selectedOption={selectedOption}
        onChange={mockChangeValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("input").getAttribute("value")).toBe("Banana");

    fireEvent.change(getByTestId("input"), { target: { value: "Dates" } });
    expect(mockChangeValue).toBeCalled();
  });

  it("should show dropdown list", () => {
    const { getByTestId, getAllByTestId, asFragment } = render(
      <Dropdown
        options={options}
        selectedOption={selectedOption}
        onChange={mockChangeValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByTestId("input-container"));
    expect(getByTestId("dropdown-list")).toHaveClass("visible");
    expect(getAllByTestId("dropdown-item").length).toBeGreaterThan(1);

    fireEvent.click(getAllByTestId("dropdown-item")[0]);
    expect(getByTestId("input").getAttribute("value")).toBe("Apple");
    expect(
      getByTestId("dropdown-container").getElementsByClassName("dropdown-item")
        .length
    ).toEqual(0);
  });
});
