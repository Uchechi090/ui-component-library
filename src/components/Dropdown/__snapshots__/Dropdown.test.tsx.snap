// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should show dropdown list 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container"
    data-testid="dropdown-container"
  >
    <div
      class="input-container"
      data-testid="input-container"
    >
      <input
        data-testid="input"
        placeholder="Select an option"
        type="text"
        value="Banana"
      />
      <div
        class="input-arrow-container"
      >
        <svg
          class="input-arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Dropdown /> should show props 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container"
    data-testid="dropdown-container"
  >
    <div
      class="input-container"
      data-testid="input-container"
    >
      <input
        data-testid="input"
        placeholder="Select an option"
        type="text"
        value="Banana"
      />
      <div
        class="input-arrow-container"
      >
        <svg
          class="input-arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
